heartbeat/scheduler/scheduler.go

```diff
@@ -220,8 +223,22 @@ func (s *Scheduler) runOnce(runAt time.Time, taskFn timerqueue.TimerTaskFn) {
 
 	// Schedule task to run sometime in the future. Wrap the task in a go-routine so it doesn't
 	// block the timer thread.
-	asyncTask := func(now time.Time) { go taskFn(now) }
-	s.timerQueue.Push(runAt, asyncTask)
+	mtx := sync.Mutex{}
+	mtx.Lock()
+	t := time.AfterFunc(runAt.Sub(now), func() {
+		taskFn()
+		mtx.Lock()
+		s.timers.Delete(id)
+		mtx.Unlock()
```

Comment:
you should not need this mutex when using `sync.Map`. On the other hand `sync.Map` adds overhead and a plain `map` with mutex might be faster.

Reply:
Yeah, it's not there to guard the actual map access, it's there to make sure that the `Store` happens before the `Delete`. AFAIK there's no guarantee that the `AfterFunc` callback won't execute before the timer is stored in the first place.

Reply:
Perhaps it'd be more efficient to have one big mutex guarding one big map, but my thinking was that since sync.Map supposedly works well when keys are not under contention having separate mutexes would be more efficient in that we wouldn't have to serialize access to the map.